src/main/java/com/coveo/saml/SamlClient.java

```diff
@@ -427,6 +432,49 @@ public static SamlClient fromMetadata(
         samlBinding);
   }
 
+  /**
+   * Wrap a {@link java.io.Reader Reader} to skip a BOM if it is present.
+   * OpenSaml won't accept a metadata file if it starts with a BOM.
+   * @param metadata The metadata with optional BOM
+   * @return A {@link BufferedReader} set to the first real character
+  */
+  private static BufferedReader skipBom(Reader metadata) throws SamlException {
+    BufferedReader bufferedMetadata = new BufferedReader(metadata);
+    try {
+      String bom = "\uFEFF";
+      char[] maybeBom = new char[1];
+      bufferedMetadata.mark(1);
+      bufferedMetadata.read(maybeBom);
+      if (!bom.equals(new String(maybeBom))) {
+        bufferedMetadata.reset();
+      }
+    } catch (IOException e) {
+      throw new SamlException("Couldn't read metadata", e);
+    }
+    return bufferedMetadata;
+  }
+
+  /**
+   * Convert a Base64 String to a Reader then inflates if necessary.
+   * @param encodedResponse a Base64 String with optionally deflated xml
+   * @return A Reader with decoded and inflated xml
+   */
+  private static Reader decodeAndInflate(String encodedResponse) {
+    ByteArrayInputStream afterB64Decode =
+        new ByteArrayInputStream(Base64.decodeBase64(encodedResponse));
+
+    char firstChar = (char) afterB64Decode.read();
+    afterB64Decode.reset();
+    Reader reader;
+    if (firstChar != '<') {
```

Comment:
I don't think this is appropriate, even though it probably works 99.9999999...% of the time. I looked (briefly) at the DEFLATE spec and there is a tiny possibility that a deflated stream could start with the bits making up that character.

From what I see in the SAML spec, the response will be deflated if it comes from the query string, and won't be otherwise. I think that adding an argument to the public calls that take a response to indicate the origin would be more appropriate.

Reply:
Good point on deflate only being used when it's sent through a query parameter. I'll add an option for passing in the request method.